core/src/main/java/com/opensymphony/xwork2/util/AnnotationUtils.java

```diff
@@ -150,25 +166,32 @@ public static void addAllInterfaces(Class clazz, List<Class> allInterfaces) {
      * @return the annotation found, or {@code null} if none
      */
     public static <A extends Annotation> A findAnnotation(Method method, Class<A> annotationType) {
-        A result = getAnnotation(method, annotationType);
-        Class<?> clazz = method.getDeclaringClass();
+        AnnotationCacheKey cacheKey = new AnnotationCacheKey(method, annotationType);
+        A result = (A) findAnnotationCache.get(cacheKey);
         if (result == null) {
-            result = searchOnInterfaces(method, annotationType, clazz.getInterfaces());
-        }
-        while (result == null) {
-            clazz = clazz.getSuperclass();
-            if (clazz == null || clazz.equals(Object.class)) {
-                break;
-            }
-            try {
-                Method equivalentMethod = clazz.getDeclaredMethod(method.getName(), method.getParameterTypes());
-                result = getAnnotation(equivalentMethod, annotationType);
-            } catch (NoSuchMethodException ex) {
-                // No equivalent method found
-            }
+            result = getAnnotation(method, annotationType);
+            Class<?> clazz = method.getDeclaringClass();
             if (result == null) {
                 result = searchOnInterfaces(method, annotationType, clazz.getInterfaces());
             }
+            while (result == null) {
+                clazz = clazz.getSuperclass();
+                if (clazz == null || clazz.equals(Object.class)) {
+                    break;
+                }
+                try {
+                    Method equivalentMethod = clazz.getDeclaredMethod(method.getName(), method.getParameterTypes());
+                    result = getAnnotation(equivalentMethod, annotationType);
+                } catch (NoSuchMethodException ex) {
+                    // No equivalent method found
```

Comment:
No, it is common to not found the method in one of supercleasses during our hierarchically search.